osm_fudge/index/bk_tree.py

```diff
@@ -0,0 +1,90 @@
+'''An implementation of bk-tree
+
+More: https://en.wikipedia.org/wiki/BK-tree
+'''
+
+
+def grouped(iterable, n):
+    return zip(*[iter(iterable)]*n)
+
+
+class BKTree:
+    '''A bk-tree implementation and defined user methods
+       for insertion and querying the string
+    '''
+
+    def __init__(self, distanceMetrics):
+        self.tree = [[]]
+        self.distanceMetrics = distanceMetrics
+
+    def insert_bk_element(self, level_index, str, dist):
```

Comment:
Hm this is getting a bit complicated. What about making the tree explicit roughly like the following pseudo-code:

```python
class Node: # this class is an impl. detail of the bk-tree; could be inside the bk-tree class or append `_` to make clear to user not to use
    def __init__(self, value, edges, children):
        # store; edges is list of distance metric number, children is list of Nodes; len(edges) == len(children), parallel arrays

    def sort(self):
        # sort edges and children parallel arrays by edges, for binary search

class BKTree:
    def __init__(self, metric):
        self.metric = metric
        self.root = None

    def sort(self):
        # recursively calls node.sort() on all tree nodes,
        # then set self.sorted=True; lookup with binary search can then assert that all nodes are sorted

    def insert(self, value):
        if not root:
            root = Node(value, [] [])
        else:
            # recursively descend based on parallel arrays root.edges, root.children, etc.
            # and insert at correct position: compute distance with self.metric and add to self.edges, add sub-tree to self.children

    def nearest(self, value, max_distance=None, max_results=None):
        assert self.sorted
        # another option is to fall back to a slow linear scan when !sorted, and a binary search when sorted
```

You could look into `namedtuple`s instead of classes, also maybe call the bk-tree public member function the user has to call before lookups can take place `freeze()`?